### PR TITLE
Willr/feature/drivertrim

### PR DESCRIPTION
**Pull request by drhieu**
Updated to previous PR(closed)
+added driver trim
+added turn damping at high speed
+added "haptic" feedback
+added scale buttons
+fixed duty cycle clamped issue

### CODE REVIEW

File: rr_rover_zero_driver/scripts/rover_zero.py [hunk 1]
```diff
@@ -402,6 +431,16 @@ def get_motor_current(self):
...
             self._right_motor_current = m2_current / 100.0
 
+    def get_motor_speed(self):
+        self._serial_lock.acquire()
+        (res1, m1_speed) = self._roboclaw.ReadSpeedM1(self._address)
+        (res2, m2_speed) = self._roboclaw.ReadSpeedM2(self._address)
+        self._serial_lock.release()
+
+        if res1 and res2:
+            self._left_motor_speed = m1_speed / 92
```
Comment (jbrover) [inline rr_rover_zero_driver/scripts/rover_zero.py]: What is the 92? Seems like a magic number.

File: rr_rover_zero_driver/scripts/rover_zero.py [hunk 2]
```diff
@@ -402,6 +431,16 @@ def get_motor_current(self):
             self._left_motor_current = m1_current / 100.0
```
Comment (jbrover) [inline rr_rover_zero_driver/scripts/rover_zero.py]: Is the 100 another magic number?

File: rr_rover_zero_driver/scripts/rover_zero.py [hunk 3]
```diff
@@ -208,14 +217,16 @@ def verify_ros_parameters(self):
 
         if self._active_brake_timeout <= 0:
             rospy.logwarn('Active breaking timeout (active_brake_timeout) is disabled. If PID control is enabled and active breaking is disabled when idle, passive current may be applied motors causing potential motor overheating or reduced battery life.')
+        if not self._odom_frequency <= 0:
```
Comment (jbrover) [inline rr_rover_zero_driver/scripts/rover_zero.py]: The 'not' here seems suspicious. Is this a mistake??

File: rr_rover_zero_driver/launch/teleop.launch [hunk 4]
```diff
@@ -38,20 +41,21 @@
...
-        <param name="y_button_toggle" value="true"/>
+    <!-- PS4 Ubuntu Driver -->
+    <node pkg="ds4_driver" type="ds4_driver_node.py" name="ds4_driver" output="screen" >
+	<rosparam command="load" file="$(find rr_rover_zero_driver)/config/ps4_driver_config.yaml" />
+	<param name="device_addr" value="" />
+	<param name="use_standard_msgs" value="false" />
+	<param name="autorepeat_rate" value="0" />
+    </node>
+
+    <node pkg="tf2_ros" type="static_transform_publisher" name="ds4_to_imu"
```
Comment (jbrover) [inline rr_rover_zero_driver/launch/teleop.launch]: What does this do?

Comment (drhieu) [reply rr_rover_zero_driver/launch/teleop.launch]: its an artifacts from testing. removed

File: rr_rover_zero_driver/scripts/rover_zero.py (hunk 2 above)
Comment (drhieu) [reply rr_rover_zero_driver/scripts/rover_zero.py]: dividing by 100 will give you the current in amp (https://downloads.basicmicro.com/docs/roboclaw_user_manual.pdf) SEARCH FOR Address, 49

File: rr_rover_zero_driver/scripts/rover_zero.py (hunk 1 above)
Comment (drhieu) [reply rr_rover_zero_driver/scripts/rover_zero.py]: updated

File: rr_rover_zero_driver/scripts/rover_zero.py (hunk 3 above)
Comment (drhieu) [reply rr_rover_zero_driver/scripts/rover_zero.py]: yup; fixed

**Review by jbrover — APPROVED** [review]
comments resolved. approved.

**Pull request merged**